### PR TITLE
EVP_MAC XOF proposal

### LINKED ISSUE #13232: EVP_MAC does not document how to use it in XOF mode
We have at least one MAC in the EVP_MAC interface that support XOF, but it's unclear to me how you can actually use it in XOF mode. For an EVP_MD we have EVP_DigestFinalXOF().

(It also doesn't make sense to me that you need to set an xof parameter to enable the xof mode.)

### LINKED ISSUE #14140: Proposal required for how to use XOF mode for EVP_MAC
In order to make progress on issue #13232 we need a proposal for how XOF mode should be used in conjunction with EVP_MAC

### PR DESCRIPTION
**Pull request by paulidale**
This PR adds a proposed `EVP_MAC_finalXOF()` function.

This function sets the XOF flag for the MAC and then calls its final.

Calling this final function multiple times is _not_ supported currently.  Doing that entails significant rework to the underlying KECCAK/SHAKE code and the OTC decided that this was beyond scope for 3.0.

Additional error handling is added to `EVP_MAC_final()`.

- [x] documentation is added or updated
- [x] tests are added or updated

Fixes #14140 by being the proposal
Fixes #13232 by adding documentation

### CODE REVIEW

File: crypto/err/openssl.txt [hunk 1]
```diff
@@ -689,6 +689,7 @@ EVP_R_INVALID_KEY:163:invalid key
 EVP_R_INVALID_KEY_LENGTH:130:invalid key length
 EVP_R_INVALID_LENGTH:221:invalid length
 EVP_R_INVALID_NULL_ALGORITHM:218:invalid null algorithm
+EVP_R_INVALID_NULL_ARGUMENT:226:invalid null argument
```
Comment (levitte) [inline crypto/err/openssl.txt:692]: Please use `ERR_R_PASSED_NULL_PARAMETER`

Comment (paulidale) [reply crypto/err/openssl.txt]: I thought there was an error for this but didn't find it.  Have changed.

**Comment by paulidale** [comment]
Is an OTC vote/discussion required for this?

As noted above, the decisions that multiple final calls are not going to be supported in 3.0 has been made for digests and that directly flows through to here.

**Comment by paulidale** [comment]
With the OTC decision made, review please?

File: crypto/evp/mac_lib.c [hunk 2]
```diff
@@ -116,21 +116,56 @@ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen)
...
+    if (out == NULL) {
+        if (outl == NULL) {
+            ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_NULL_PARAMETER);
+            return 0;
+        }
+        *outl = EVP_MAC_CTX_get_mac_size(ctx);
+        return 1;
+    }
+    if (xof) {
+        params[0] = OSSL_PARAM_construct_int(OSSL_DIGEST_PARAM_XOFLEN, &xof);
```
Comment (mattcaswell) [inline crypto/evp/mac_lib.c:144]: Shouldn't this be `&outsize`? If so how does this pass tests?

File: test/evp_test.c [hunk 3]
```diff
@@ -1360,20 +1368,31 @@ static int mac_test_run_mac(EVP_TEST *t)
...
-        goto err;
+    if (expected->xof) {
+        if (!TEST_ptr(got = OPENSSL_malloc(expected->output_len))) {
+            t->err = "TEST_FAILURE";
+            goto err;
+        }
+        if (!EVP_MAC_finalXOF(ctx, got, expected->output_len)) {
+            t->err = "MAC_FINAL_ERROR";
+            goto err;
+        }
```
Comment (mattcaswell) [inline test/evp_test.c:1379]: Doesn't this need to do the `memory_err_compare` call like in the other branch of the if? (Possibly the reason the tests are passing)

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/evp/mac_lib.c (hunk 2 above)
Comment (paulidale) [reply crypto/evp/mac_lib.c:144]: No, it should be `OSSL_MAC_PARAM_XOF`.  KMAC takes a flag to set XOF mode not an XOF length.

File: test/evp_test.c (hunk 3 above)
Comment (paulidale) [reply test/evp_test.c:1379]: Yep :(

**Comment by paulidale** [comment]
Fixes pushed.

**Review by mattcaswell — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Merged to master after removing some trailing whitespace.

**Pull request closed without merging**